File: Dsa-day20.cpp
class SegmentTree {
public:
    vector<int> tree;
    int n;

    SegmentTree(vector<int>& baskets) {
        n = baskets.size();
        tree.resize(4 * n);
        build(baskets, 0, 0, n - 1);
    }

    void build(vector<int>& baskets, int node, int l, int r) {
        if (l == r) {
            tree[node] = baskets[l];
            return;
        }
        int mid = (l + r) / 2;
        build(baskets, 2 * node + 1, l, mid);
        build(baskets, 2 * node + 2, mid + 1, r);
        tree[node] = max(tree[2 * node + 1], tree[2 * node + 2]);
    }

    bool placeFruit(int node, int l, int r, int fruit) {
        if (tree[node] < fruit) return false;

        if (l == r) {
            tree[node] = 0;
            return true;
        }

        int mid = (l + r) / 2;
        bool placed = placeFruit(2 * node + 1, l, mid, fruit);
        if (!placed)
            placed = placeFruit(2 * node + 2, mid + 1, r, fruit);

        tree[node] = max(tree[2 * node + 1], tree[2 * node + 2]);
        return placed;
    }
};

class Solution {
public:
    int numOfUnplacedFruits(vector<int>& fruits, vector<int>& baskets) {
        SegmentTree seg(baskets);
        int unplaced = 0;

        for (int fruit : fruits) {
            if (!seg.placeFruit(0, 0, baskets.size() - 1, fruit)) {
                unplaced++;
            }
        }

        return unplaced;
    }
};

File: Dsa-day20.py
class Solution:
    def numOfUnplacedFruits(self, fruits: List[int], baskets: List[int]) -> int:
        class Tree:
            def __init__(self, val = None):
                self.val = val
                self.left = None
                self.right = None

        def visitor(node):
            return node.left.val if node.left.val >= node.right.val else node.right.val

        def segment_tree(baskets):
            def dfs(root, start, end):
                mid = (start + end) // 2
                if start < mid:
                    root.left = Tree(0)
                    dfs(root.left, start, mid)
                    root.left.val = visitor(root.left)
                else:
                    root.left = Tree(baskets[start])
                if mid + 1 < end:
                    root.right = Tree(0)
                    dfs(root.right, mid + 1, end)
                    root.right.val = visitor(root.right)
                else:
                    root.right = Tree(baskets[end])
            root = Tree(0)
            dfs(root, 0, len(baskets) - 1)
            root.val = visitor(root)
            return root

        def is_placed(root, fruit):
            if fruit > root.val:
                return False
            if root.left and fruit <= root.left.val:
                is_placed(root.left, fruit)
            elif root.right:
                is_placed(root.right, fruit)
            if not root.left and not root.right:
                root.val = 0
            else:
                root.val = visitor(root)
            return True

        unplaced = len(fruits)
        root = segment_tree(baskets)
        for fruit in fruits:
            if is_placed(root, fruit):
                unplaced -= 1
        return unplaced